[Clang] Consider reachability for file-scope warnings on initializers

JustinStitt opened the pull request:
Every once in awhile, we get a new Linux kernel report about false positive clang warnings being emitted for file-scope variables when initialized. See [[0]](https://lore.kernel.org/all/8c252429-5000-0649-c49f-8225d911241b@gmail.com/) and [[1]](https://lore.kernel.org/all/202112181827.o3X7GmHz-lkp@intel.com/).

To help fix this, back in 2019 Nathan Huckleberry sent a phab PR [D63889](https://reviews.llvm.org/D63889). Nathan was @nickdesaulniers' intern and most likely had to return to school before that PR landed. I was able to use a good amount of code from that PR with a few workarounds to fix bitrot and clang regression suite failures. So, consider this a spiritual revival of that patchset.

cc @nickdesaulniers @kees @nathanchance (the other Nathan)


---

Suppress runtime warnings for unreachable code in global variable initializers while maintaining warnings for the same code in functions.

For context, some global variable declarations using ternaries can emit warnings during initialization even when a particular expression is never used. This behavior differs from GCC since they properly emit warnings based on reachability -- even at file scope.

The simplest example is:
```c
$ cat test.c
unsigned long long b1 = 1 ? 0 : 1ULL << 64;

$ clang-21 -fsyntax-only test.c
test.c:1:39: warning: shift count >= width of type [-Wshift-count-overflow]
    1 | unsigned long long foo = 1 ? 0 : 1ULL << 64;
      |                                       ^  ~~
$ gcc-13 -fsyntax-only test.c
<empty>
```

Clang previously emitted a ``-Wshift-count-overflow`` warning regardless of branch taken. This PR constructs a CFG and defers runtime diagnostic emission until we can analyze the CFG for reachability.

To be clear, the intention is only to modify initializers in global scope and only remove warnings if unreachable, no new warnings should pop up anywhere.

Link: https://reviews.llvm.org/D63889 (original PR by Nathan Huckleberry)
Fixes: https://github.com/clangBuiltLinux/linux/issues/92
Fixes: https://github.com/llvm/llvm-project/issues/38137

JustinStitt committed:
[Clang] Consider reachability for file-scope warnings on initializers

Suppress runtime warnings for unreachable code in global variable
initializers while maintaining warnings for the same code in functions.

For context, some global variable declarations using ternaries can emit
warnings during initialization even when a particular expression is
never used. This behavior differs from GCC since they properly emit
warnings based on reachability -- even at file scope.

The simplest example is:
```c
$ cat test.c
unsigned long long b1 = 1 ? 0 : 1ULL << 64;

$ clang-21 -fsyntax-only test.c
test.c:1:39: warning: shift count >= width of type [-Wshift-count-overflow]
    1 | unsigned long long foo = 1 ? 0 : 1ULL << 64;
      |                                       ^  ~~
$ gcc-13 -fsyntax-only test.c
<empty>
```

Clang previously emitted a ``-Wshift-count-overflow`` warning regardless
of branch taken. This PR constructs a CFG and defers runtime diagnostic
emission until we can analyze the CFG for reachability.

To be clear, the intention is only to modify initializers in global
scope and only remove warnings if unreachable, no new warnings should
pop up anywhere.

Link: https://reviews.llvm.org/D63889 (original PR by Nathan Huckleberry)
Link: https://github.com/clangBuiltLinux/linux/issues/92
Co-Authored-By: Nathan Huckleberry <nhuck@google.com>
Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt commented:
also Fixes: https://github.com/llvm/llvm-project/issues/38137

JustinStitt committed:
remove fixme

Signed-off-by: Justin Stitt <justinstitt@google.com>

zwuis commented on clang/include/clang/Parse/Parser.h:
```diff
@@ -5223,10 +5223,31 @@ class Parser : public CodeCompletionHandler {
+    }
+
+    return init;
   }
```
Do we need to move the implementation to source file(.cpp)?

zwuis commented on clang/lib/Sema/AnalysisBasedWarnings.cpp:
```diff
@@ -2724,6 +2724,80 @@ static void flushDiagnostics(Sema &S, const sema::FunctionScopeInfo *fscope) {
+        // reachable.
+      }
+
+      if (AllReachable)
```
While we are here, can we use `llvm::all_of`?

zwuis commented on clang/lib/Sema/AnalysisBasedWarnings.cpp:
```diff
@@ -2724,6 +2724,80 @@ static void flushDiagnostics(Sema &S, const sema::FunctionScopeInfo *fscope) {
+  if (VarDeclPossiblyUnreachableDiags.find(VD) ==
+      VarDeclPossiblyUnreachableDiags.end()) {
+    return;
+  }
```
```cpp
if (...)
  return;
```

Can we use `llvm::is_contained`?

zwuis commented on clang/lib/Sema/AnalysisBasedWarnings.cpp:
```diff
@@ -2724,6 +2724,80 @@ static void flushDiagnostics(Sema &S, const sema::FunctionScopeInfo *fscope) {
+    return;
+  }
+
+  AnalysisDeclContext AC(nullptr, VD);
```
```cpp
AnalysisDeclContext AC(/*Mgr=*/nullptr, VD);
```

zwuis reviewed (commented):
Thanks! Please add a release note.

JustinStitt committed:
various fixes and refactors

based on review from gh@zwuis

- add release note
- move Parser::ParseInitializer impl to a source file
- remove FlushDiagnostics, favoring open-coded version
- DeclForInitializer is now a VarDecl in Sema.h
- use certain llvm helpers like all_of and is_contained

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt committed:
dont recompute Analysis in llvm::all_of

Signed-off-by: Justin Stitt <justinstitt@google.com>

cor3ntin commented on clang/include/clang/Sema/AnalysisBasedWarnings.h:
```diff
@@ -57,6 +61,8 @@ class AnalysisBasedWarnings {
   enum VisitFlag { NotVisited = 0, Visited = 1, Pending = 2 };
   llvm::DenseMap<const FunctionDecl*, VisitFlag> VisitedFD;
+  llvm::MapVector<VarDecl *, SmallVector<PossiblyUnreachableDiag, 4>>
+      VarDeclPossiblyUnreachableDiags;
```
I feel like a `std::multimap<VarDecl *,  PossiblyUnreachableDiag>` would be a more suited container here

(still preserve order and avoid having lots of SmallVector)

JustinStitt replied:
Hi, this is done in [`c6ec954` (#163885)](https://github.com/llvm/llvm-project/pull/163885/commits/c6ec954bf777b812848774baf61f0e74a922d1d1) but I think it wasn't as elegant as I was expecting. This is because `SmallVector<PossiblyUnreachableDiag, 4> PossiblyUnreachableDiags;` in `ScopeInfo.h` is already declared as a vector. The change meant I needed to template `emitPossiblyUnreachableDiags` and use `llvm::make_second_range`.

cor3ntin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -6750,6 +6750,11 @@ class Sema final : public SemaBase {
+    /// Declaration for initializer if one is currently being
+    /// parsed. Used when an expression has a possibly unreachable
+    /// diagnostic to reference the declaration as a whole.
+    VarDecl *DeclForInitializer = nullptr;
```
Would that be ever different from ManglingContextDecl? Can we reuse that?

cor3ntin commented on clang/lib/Parse/ParseInit.cpp:
```diff
@@ -581,3 +581,28 @@ bool Parser::ParseMicrosoftIfExistsBraceInitializer(ExprVector &InitExprs,
+        VD && VD->isFileVarDecl()) {
+      if (!VD->getType()->isReferenceType() || VD->isConstexpr()) {
+        Actions.ExprEvalContexts.back().DeclForInitializer = VD;
+      }
```
```suggestion
    if (auto *VD = dyn_cast<VarDecl>(DeclForInitializer);
        VD && VD->isFileVarDecl() && (!VD->getType()->isReferenceType() || VD->isConstexpr()))
        Actions.ExprEvalContexts.back().DeclForInitializer = VD;
```

cor3ntin commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -13727,6 +13734,16 @@ void Sema::DiagnoseUniqueObjectDuplication(const VarDecl *VD) {
+      if (!S.ExprEvalContexts.empty())
+        S.ExprEvalContexts.back().DeclForInitializer = nullptr;
+    }
+  } Clearer(*this);
```
It would be simpler to use `make_scope_exit` here

JustinStitt committed:
use multimap over MapVector

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt committed:
various changes from review

based on review from gh@zwuis and gh@cor3ntin

- collapse if into parent if
- use make_scope_exit over hand-rolled RAII obj
- avoid casting a DeclForInitializer back to VarDecl since it already is
  one
- make a few functions correctly formatted with lowercase
- make EmitPossiblyUnreachableDiags static

zwuis commented on clang/lib/Sema/AnalysisBasedWarnings.cpp:
```diff
@@ -2935,45 +2991,7 @@ void clang::sema::AnalysisBasedWarnings::IssueWarnings(
-          const CFGBlock *block = AC.getBlockForRegisteredExpression(S);
-          CFGReverseBlockReachabilityAnalysis *cra =
-              AC.getCFGReachablityAnalysis();
-          // FIXME: We should be able to assert that block is non-null, but
```
I have no idea about this but I think we should keep this comment because it's not fixed.

JustinStitt committed:
reinstate FIXME as it was not fixed

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt commented:
ping

kees reviewed (approved):
This behaves how I am expecting it to. Thank you for tackling this!

erichkeane reviewed (commented):
It is INCREDIBLY unfortunate that this got merged with open Parse-maintainer objections, and happened almost exclusively while half the maintainers were away for LLVM-Dev and WG21 meetings.  

The `DeclForInitializer` part seems very incorrect to me (though the parser maintainers  should better look into this), and doesn't seem like it'll properly work in dependent cases.

I also don't really see any feedback from our CFG code owners either.


IMO, this was merged prematurely.

erichkeane commented:
> Shoot, sorry about the bad timing. Do your concerns warrant a revert? Is there anyone I should manually add as a reviewer? @erichkeane

I don't think so at the moment.  I'd like to see Corentin take a run at this again, as this is mostly in his backyard, but he's still away on the WG21 trip for the next few days.


> AFAICT, the only open "objection" is regarding a data structure choice. I switched over to @cor3ntin's recommendation but there was some minor nuance that I tried to get more feedback on.

I see, I am perhaps less concerned there then.  At the moment, I'd hold off on a revert, and we'll see what Corentin has to say when he gets back.

nickdesaulniers commented:
> and happened almost exclusively while half the maintainers were away for LLVM-Dev 

Guys, we were _all_ at LLVM-Dev _together_.  Next time, let me make some introductions. :P
